File: token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
}

const (
	// utility tokens
	Illegal = "ILLEGAL"
	EOF     = "EOF"

	// identifiers & literals
	Ident    = "IDENTIFIER"
	Integer  = "INTEGER"
	Float    = "FLOAT"
	SQString = "STRING(SQ)"
	DQString = "STRING(DQ)"

	// keywords
	Abstract    = "abstract"
	And         = "and"
	Array       = "array"
	As          = "as"
	Break       = "break"
	Callable    = "callable"
	Case        = "case"
	Catch       = "catch"
	Class       = "class"
	Clone       = "clone"
	Const       = "const"
	Continue    = "continue"
	Declare     = "declare"
	Default     = "default"
	Die         = "die"
	Do          = "do"
	Echo        = "echo"
	Else        = "else"
	ElseIf      = "elseif"
	Empty       = "empty"
	EndDeclare  = "enddeclare"
	EndFor      = "endfor"
	EndForEach  = "endforeach"
	EndIf       = "endif"
	EndSwitch   = "endswitch"
	EndWhile    = "endwhile"
	Eval        = "eval"
	Exit        = "exit"
	Extends     = "extends"
	Final       = "final"
	Finally     = "finally"
	For         = "for"
	ForEach     = "foreach"
	Funtion     = "function"
	Global      = "global"
	Goto        = "goto"
	If          = "if"
	Implements  = "implements"
	Include     = "include"
	IncludeOnce = "include_once"
	InstanceOf  = "instanceof"
	InsteadOf   = "insteadof"
	Interface   = "interface"
	IsSet       = "isset"
	List        = "list"
	Namespace   = "namespace"
	New         = "new"
	Or          = "or"
	Print       = "print"
	Private     = "private"
	Protected   = "protected"
	Public      = "public"
	Require     = "require"
	RequireOnce = "require_once"
	Return      = "return"
	Static      = "static"
	Switch      = "switch"
	Throw       = "throw"
	Trait       = "trait"
	Try         = "try"
	Unset       = "unset"
	Use         = "use"
	Var         = "var"
	While       = "while"
	Xor         = "xor"
	Yield       = "yield"
	YieldFrom   = "yield from"

	// one-character punctuators
	LSquare = "["
	RSquare = "]"
	LParen  = "("
	RParen  = ")"
	LBrace  = "{"
	RBrace  = "}"

	// one-character operators
	Dot      = "."
	Plus     = "+"
	Dash     = "-"
	Star     = "*"
	Tilde    = "~"
	Bang     = "!"
	Dollar   = "$"
	FSlash   = "/"
	BSlash   = "\\"
	Percent  = "%"
	Less     = "<"
	More     = ">"
	Eq       = "="
	Caret    = "^"
	Pipe     = "|"
	Amper    = "&"
	Question = "?"
	Colon    = ":"
	Semi     = ";"
	Comma    = ","

	// two-character operators
	Arrow       = "->"
	TwoPlus     = "++"
	TwoDash     = "--"
	TwoStar     = "**"
	TwoLess     = "<<"
	TwoMore     = ">>"
	LessEq      = "<="
	MoreEq      = ">="
	TwoEq       = "=="
	BangEq      = "!="
	TwoPipe     = "||"
	TwoAmper    = "&&"
	StarEq      = "*="
	FSlashEq    = "/="
	PercentEq   = "%="
	PlusEq      = "+="
	DashEq      = "-="
	DotEq       = ".="
	AmperEq     = "&="
	CaretEq     = "^="
	PipeEq      = "|="
	TwoQuestion = "??"

	// three-character operators
	ThreeEq    = "==="
	BangTwoEq  = "!=="
	TwoStarEq  = "**="
	TwoLessEq  = "<<="
	ThreeLess  = "<<<"
	TwoMoreEq  = ">>="
	LessEqMore = "<=>"
	Spread     = "..."
	// NOTE: I think `??=` is missing from this
)

File: src/lexer/token.rs
use std::fmt::{Display, Formatter, Result};

#[derive(Debug)]
pub enum TokenKind {
    // utility tokens
    Illegal(String),
    Comment,

    // identifiers & literals
    Ident(String),
    Integer(String),
    Float(String),
    SQString(String),
    DQString(String),
    HDString(String),
    NDString(String),

    // keywords
    Abstract,
    And,
    Array,
    As,
    Break,
    Callable,
    Case,
    Catch,
    Class,
    Clone,
    Const,
    Continue,
    Declare,
    Default,
    Die,
    Do,
    Echo,
    Else,
    ElseIf,
    Empty,
    EndDeclare,
    EndFor,
    EndForEach,
    EndIf,
    EndSwitch,
    EndWhile,
    Eval,
    Exit,
    Extends,
    Final,
    Finally,
    For,
    ForEach,
    Funtion,
    Global,
    Goto,
    If,
    Implements,
    Include,
    IncludeOnce,
    InstanceOf,
    InsteadOf,
    Interface,
    IsSet,
    List,
    Namespace,
    New,
    Or,
    Print,
    Private,
    Protected,
    Public,
    Readonly,
    Require,
    RequireOnce,
    Return,
    Static,
    Switch,
    Throw,
    Trait,
    Try,
    Unset,
    Use,
    Var,
    While,
    Xor,
    Yield,
    YieldFrom,

    // one-character tokens
    LSquare,
    RSquare,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Dot,
    Plus,
    Dash,
    Star,
    Tilde,
    Bang,
    Dollar,
    FSlash,
    BSlash,
    Percent,
    Less,
    More,
    Eq,
    Caret,
    Pipe,
    Amper,
    Question,
    Colon,
    Semi,
    Comma,

    // two-character tokens
    Arrow,
    TwoPlus,
    TwoDash,
    TwoStar,
    TwoLess,
    TwoMore,
    LessEq,
    MoreEq,
    TwoEq,
    BangEq,
    TwoPipe,
    TwoAmper,
    StarEq,
    FSlashEq,
    PercentEq,
    PlusEq,
    DashEq,
    DotEq,
    AmperEq,
    CaretEq,
    PipeEq,
    TwoQuestion,
    QuestionMore,
    PoundLSquare,

    // three-character tokens
    ThreeEq,
    BangTwoEq,
    TwoStarEq,
    EchoOpen,
    TwoLessEq,
    ThreeLess,
    TwoMoreEq,
    LessEqMore,
    TwoQuestionEq,
    ThreeDot,

    // many-character tokens
    Open,
}

pub struct Token {
    kind: TokenKind,
    start: (usize, usize),
    end: (usize, usize),
}

impl Token {
    pub fn emit() -> Token {
        Token {
            kind: TokenKind::Illegal(String::from("test")),
            start: (0, 0),
            end: (0, 0),
        }
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut Formatter) -> Result {
        write!(
            f,
            "{:?}, [{}:{}] - [{}:{}]",
            self.kind, self.start.0, self.start.1, self.end.0, self.end.1
        )
    }
}
